finality-aleph/src/network/substrate.rs

```diff
@@ -60,163 +83,230 @@ impl ProtocolNaming {
-                                iter::once(MultiaddressProtocol::P2p(remote.into())).collect();
-                            trace!(target: "aleph-network", "Connected event from address {:?}", multiaddress);
-                            if let Err(e) = self.network.add_peers_to_reserved_set(
-                                self.naming.protocol_name(&Protocol::Authentication),
-                                iter::once(multiaddress.clone()).collect(),
-                            ) {
-                                error!(target: "aleph-network", "add_reserved failed for authentications: {}", e);
-                            }
-                            if let Err(e) = self.network.add_peers_to_reserved_set(
-                                self.naming.protocol_name(&Protocol::BlockSync),
-                                iter::once(multiaddress).collect(),
-                            ) {
-                                error!(target: "aleph-network", "add_reserved failed for block sync: {}", e);
-                            }
-                            continue;
-                        }
-                        PeerDisconnected(remote) => {
-                            trace!(target: "aleph-network", "Disconnected event for peer {:?}", remote);
-                            let addresses: Vec<_> = iter::once(remote).collect();
-                            if let Err(e) = self.network.remove_peers_from_reserved_set(
-                                self.naming.protocol_name(&Protocol::Authentication),
-                                addresses.clone(),
-                            ) {
-                                warn!(target: "aleph-network", "Error while removing peer from Protocol::Authentication reserved set: {}", e)
-                            }
-                            if let Err(e) = self.network.remove_peers_from_reserved_set(
-                                self.naming.protocol_name(&Protocol::BlockSync),
-                                addresses,
-                            ) {
-                                warn!(target: "aleph-network", "Error while removing peer from Protocol::BlockSync reserved set: {}", e)
-                            }
-                            continue;
+impl NotificationService {
+    pub fn new(service: Box<dyn sc_network::config::NotificationService>) -> Self {
+        Self {
+            service,
+            connected_peers: HashSet::new(),
+            last_status_report: time::Instant::now(),
+        }
+    }
+
+    fn random_peer<'a>(&'a self, peer_ids: &'a HashSet<PeerId>) -> Option<&'a PeerId> {
+        peer_ids
+            .intersection(&self.connected_peers)
+            .choose(&mut thread_rng())
+            .or_else(|| self.connected_peers.iter().choose(&mut thread_rng()))
+    }
+
+    fn handle_network_event(&mut self, event: SubstrateEvent) -> Option<(Vec<u8>, PeerId)> {
+        use SubstrateEvent::*;
+        match event {
+            ValidateInboundSubstream {
```

Comment:
Darn, it just occurred to me that we could simplify compatibility changes in the network massively by using these handshakes. Very out of scope for this PR, but perhaps worth creating a task.

Reply:
Added A0-4292 then.